### ISSUE TITLE
[Release-1.25] - Use base template in etc/containerd/config.toml.tmpl

### ISSUE THREAD

**Issue by brandond** [k3s-io/k3s#8117]
Backport fix for Use base template in etc/containerd/config.toml.tmpl

* #7992

**Comment by bguzman-3pillar** [comment]
### Validated on 
```
k3s -v
k3s version v1.25.12+k3s-8e945c53 (8e945c53)
go version go1.20.6
```
 
## Environment Details

 
*Infrastructure*
- [ ] Cloud
- [ ] Hosted
 
*Node(s) CPU architecture, OS, and Version:*

```
cat /etc/os-release
NAME="Red Hat Enterprise Linux"
VERSION="8.7 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.7"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.7 (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8::baseos"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/red_hat_enterprise_linux/8/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.7
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.7"
```
 
*Cluster Configuration:*

```
1 server
```
 
*Config.yaml:*

```
cat /etc/rancher/k3s/config.yaml
selinux: true
```
 
*Additional files*

```
#/var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl

{{ template "base" . }}

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."custom"]
  runtime_type = "io.containerd.runc.v2"
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."custom".options]
  BinaryName = "/usr/bin/custom-container-runtime"
```
 
## Testing Steps

1. Copy config.yaml
```
$ sudo mkdir -p /etc/rancher/k3s && sudo cp config.yaml /etc/rancher/k3s
```
2. Copy `config.toml.tmpl` under `/var/lib/rancher/k3s/agent/etc/containerd`
```
sudo cp config.toml.tmpl /var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl
```
3. Install k3s
4. `sudo cat /var/lib/rancher/k3s/agent/etc/containerd/config.toml` new base should be there
5. Cluster is up and running
 
 
 
**Validation Results:**

```
sudo cat /var/lib/rancher/k3s/agent/etc/containerd/config.toml

# File generated by k3s. DO NOT EDIT. Use config.toml.tmpl instead.
version = 2

[plugins."io.containerd.internal.v1.opt"]
  path = "/var/lib/rancher/k3s/agent/containerd"
[plugins."io.containerd.grpc.v1.cri"]
  stream_server_address = "127.0.0.1"
  stream_server_port = "10010"
  enable_selinux = true
  enable_unprivileged_ports = true
  enable_unprivileged_icmp = true
  sandbox_image = "rancher/mirrored-pause:3.6"

[plugins."io.containerd.grpc.v1.cri".containerd]
  snapshotter = "overlayfs"
  disable_snapshot_annotations = true


[plugins."io.containerd.grpc.v1.cri".cni]
  bin_dir = "/var/lib/rancher/k3s/data/82f2dc0065bdc6100f52c8365e37f8bab57f1d27c491fb0d554c0bc7bdea886a/bin"
  conf_dir = "/var/lib/rancher/k3s/agent/etc/cni/net.d"


[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
  runtime_type = "io.containerd.runc.v2"

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
  SystemdCgroup = true











[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."custom"]
  runtime_type = "io.containerd.runc.v2"
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."custom".options]
  BinaryName = "/usr/bin/custom-container-runtime"
```

```
kubectl get node,pod -A
NAME                                              STATUS   ROLES                  AGE     VERSION
node/ip-172-31-21-81.us-east-2.compute.internal   Ready    control-plane,master   4m26s   v1.25.12+k3s-8e945c53

NAMESPACE     NAME                                          READY   STATUS      RESTARTS   AGE
kube-system   pod/local-path-provisioner-69dff9496c-2ssvh   1/1     Running     0          4m11s
kube-system   pod/coredns-8b9777675-q59d8                   1/1     Running     0          4m11s
kube-system   pod/helm-install-traefik-crd-bmrs7            0/1     Completed   0          4m12s
kube-system   pod/svclb-traefik-ddc29ee2-vnr2h              2/2     Running     0          3m54s
kube-system   pod/helm-install-traefik-4ws5j                0/1     Completed   1          4m12s
kube-system   pod/traefik-66fd46ccd-g5jnb                   1/1     Running     0          3m54s
kube-system   pod/metrics-server-854c559bd-vmqvg            1/1     Running     0          4m11s
```

**Comment by nodox** [comment]
Hey this does not work. I'm currently running into this issue with k3s (version below).

The error alerts as follows. Its seems it can't identify the base template name. What is the resolution to this given that it seemed solved in the past? @bguzman-3pillar 



```sh

Mar 08 04:07:45 crimson k3s[758337]: time="2024-03-08T04:07:45Z" level=info msg="Using containerd template at /var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl"
Mar 08 04:07:45 crimson k3s[758337]: time="2024-03-08T04:07:45Z" level=fatal msg="template: compiled_template:1:12: executing \"compiled_template\" at <{{template \"base\" .}}>: template \"base\" not defined"
Mar 08 04:07:45 crimson systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
-- An ExecStart= process belonging to unit k3s.service has exited.
```

=====
k3s version v1.24.3+k3s1 (990ba0e8)
go version go1.18.1

**Comment by brandond** [comment]
> @nodox
> k3s version v1.24.3+k3s1

Please read the docs before trying to use this feature: https://docs.k3s.io/advanced?_highlight=containerd#base-template
>> Available as of September 2023 releases: v1.24.17+k3s1, v1.25.13+k3s1, v1.26.8+k3s1, v1.27.5+k3s1, v1.28.1+k3s1
